subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ProjectPathClarifyingDescriber.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts;
+
+import org.gradle.api.Describable;
+import org.gradle.util.Path;
+
+/**
+ * Static utility class containing methods for describing a project path with clarifications
+ * that may make it clearer to the reader in error messages.
+ */
+public final class ProjectPathClarifyingDescriber {
+    private static final String PROJECT_PREFIX = "project ";
+
+    private ProjectPathClarifyingDescriber() { /* not instantiable */}
+
+    /**
+     * Describes the given {@link Describable} object, clarifying when it references the root project.
+     *
+     * @param describable the object to describe
+     * @return clarified description of the object
+     */
+    public static String describe(Describable describable) {
+        return describe(describable.getDisplayName());
+    }
+
+    /**
+     * Describes the given display path, clarifying when that path references the root project.
+     *
+     * @param displayPath the path to describe
+     * @return clarified description of the path
+     */
+    public static String describe(String displayPath) {
+        String pathAlone;
+        if (displayPath.startsWith(PROJECT_PREFIX)) {
```

Comment:
While this works, I am a bit skeptical of playing with strings here. We have this data in its un-stringed form, and I think we might be able to handle this root project detection earlier

Reply:
I experimented a little with that, and wound up with ~2000 failing tests.  I don't want to fix all of those, or even analyze which of them need to be fixed.  I want to limit this change to resolution failure messages.

The implementations of `DefaultProjectComponentIdentifier` and `DefaultProjectComponentSelector` aren't very complex about how they build their display names.  While we could do an instanceof test and interpret the `ComponentIdentifier` and `ComponentSelector` types semantically, I don't think there is a need for that here.  Alternately, we could have these interfaces (or just their `DefaultProjectXYZ` implementations) each implement a new common interface and test for that.  These kind of seem like overkill for this problem.

I was kind of thinking of this type as just a Decorator around the existing project description.  This is already cleaner than some of the existing message formatting code, which is very much still string bashing, so I'm not too concerned.

Reply:
I was not proposing changing the way `ProjectComponentIdentifier/Selector` format their display names, but that we should avoid string manipulation like this if possible. 

> While we could do an instanceof test and interpret the ComponentIdentifier and ComponentSelector types semantically, I don't think there is a need for that here.

This is more so what I was proposing. Is this that much more complicated?

Reply:
It's not that it's that much more complicated to implement, it's just that `instanceof` branching is typically considered an OOP code smell to be avoided.

I think there needs to be a stronger reason to go that route here than general dislike for conditioning on String contents.  This is meant as a decoration around the existing String, so I don't see a huge problem with processing it slightly first - it still tests the extracted path properly against the root path constant.

I could be talked into moving the whole testing part into `Path` as a new `isRootPath(String)` method that handles the prefix internally, but that's new public API.